kafka-client/src/main/java/dev/responsive/kafka/internal/clients/OriginEventRecorder.java

```diff
@@ -0,0 +1,210 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import dev.responsive.kafka.api.config.ResponsiveConfig;
+import dev.responsive.kafka.internal.license.model.LicenseInfo;
+import dev.responsive.kafka.internal.license.model.UsageBasedV1;
+import dev.responsive.kafka.internal.license.server.model.OriginEventsReportRequestV1;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.time.Duration;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.function.BiConsumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.OffsetAndMetadata;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.streams.StreamsConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OriginEventRecorder implements ResponsiveConsumer.Listener {
+
+  private static final Logger LOG = LoggerFactory.getLogger(OriginEventRecorder.class);
+
+  static final String ORIGIN_EVENT_HEADER_KEY = "_roe";
+  static final byte[] ORIGIN_EVENT_MARK = new byte[] {0x1};
+
+  private final Duration reportInterval;
+  private final BiConsumer<Map<TopicPartition, Long>, String> reportOriginEvents;
+  private final String threadId;
+
+  @GuardedBy("this")
+  private final Map<TopicPartition, Long> originEvents = new HashMap<>();
+  @GuardedBy("this")
+  private long lastCommitTs = 0L;
+
+  public OriginEventRecorder(
+      final String threadId,
+      final ResponsiveConfig config,
+      LicenseInfo license
+  ) {
+    this(
+        threadId,
+        new Reporter(config, license),
+        config.getLong(ResponsiveConfig.ORIGIN_EVENT_REPORT_INTERVAL_MS_CONFIG)
+    );
+  }
+
+  @VisibleForTesting
+  OriginEventRecorder(
+      final String threadId,
+      final BiConsumer<Map<TopicPartition, Long>, String> reportOriginEvents
+  ) {
+    this(threadId, reportOriginEvents, 10_000);
+  }
+
+  private OriginEventRecorder(
+      final String threadId,
+      final BiConsumer<Map<TopicPartition, Long>, String> reportOriginEvents,
+      final long reportIntervalMs
+  ) {
+    this.threadId = threadId;
+    this.reportOriginEvents = reportOriginEvents;
+    this.reportInterval = Duration.ofMillis(reportIntervalMs);
+  }
+
+  private synchronized void inc(final TopicPartition tp) {
+    originEvents.compute(tp, (k, v) -> v == null ? 1 : v + 1);
+  }
+
+  @Override
+  public <K, V> ConsumerRecords<K, V> onPoll(final ConsumerRecords<K, V> records) {
+    for (final ConsumerRecord<K, V> record : records) {
+      final var header = record.headers().lastHeader(ORIGIN_EVENT_HEADER_KEY);
+      if (header == null) {
+        record.headers().add(ORIGIN_EVENT_HEADER_KEY, ORIGIN_EVENT_MARK);
```

Comment:
do these headers actually get propagated down to the producer when it writes? I wonder if it would make more sense to just add the headers from the producer itself to prevent the header from being removed/messed with by streams

Reply:
On a related note, don't we need to filter out changelog records? Though I suppose they are only read by the restore consumer.